File: dart/difference-of-squares/lib/difference_of_squares.dart
import "dart:math";

class DifferenceOfSquares {
  int squareOfSum(int num) {
    var list = [for (int i = 1; i <= num; i += 1) i];
    int sum = list.fold(0, (prev, element) => prev + element);
    return pow(sum, 2).toInt();
  }

  int sumOfSquares(int num) {
    var list = [for (int i = 1; i <= num; i += 1) i];
    return list.fold(0, (prev, element) => prev + pow(element, 2).toInt());
  }

  int differenceOfSquares(int num) {
    return squareOfSum(num) - sumOfSquares(num);
  }
}

File: lua/difference-of-squares/difference-of-squares.lua
local function square_of_sum(n)
  a = 0
  for i=1, n do
    a = a + i
  end

  return a^2
end

local function sum_of_squares(n)
  a = 0
  for i=1, n do
    a = a + i^2
  end

  return a
end

local function difference_of_squares(n)
  return square_of_sum(n) - sum_of_squares(n)
end

return {
  square_of_sum = square_of_sum,
  sum_of_squares = sum_of_squares,
  difference_of_squares = difference_of_squares
}
